Add a section on CSS principles. Fixes #126

dbaron opened the pull request:
This adds a somewhat substantive section on CSS to the design principles document.  It's multiple commits that are **meant to be squashed** when landed.

You should be able to [preview this here](https://dbaron.github.io/design-principles/#css) ~at some point soon, though github seems sluggish right now and it's not up yet~.

Fixes #126.

I'd like to solicit review from @tabatkins and @fantasai as well, but I can't actually mark them in reviewers.

I'm sure there's a lot that can be changed/added here -- I'd like to strike a reasonable balance between trying to polish this, and also trying to defer most "also add X" requests to later PRs.

dbaron commented:
I realize the section on computed values should probably say something about animation as well.

hober commented on index.bs:
```diff
@@ -349,6 +349,218 @@ One example of a feature that should be limited to secure contexts
 
+<h2 id="css">Cascading Style Sheets (CSS)</h2>
+
+<h3 id="css-versus-markup">Presentation should be in CSS, content should be in HTML</h3>
```
This principle strikes me as essentially the same thing as the [Separation of Concerns](https://www.w3.org/TR/html-design-principles/#separation-of-concerns) [HTML Design Principle](https://www.w3.org/TR/html-design-principles/).

If you agree, I should track this PR as part of https://github.com/w3ctag/design-principles/projects/1, and maybe the principle should be moved under [§ 2. API Design Across Languages](https://w3ctag.github.io/design-principles/#api-across-languages).

dbaron replied:
As discussed in today's teleconference, I'm going to drop this section and merge the rest.  Landing a replacement for this text is covered in #169.

tabatkins commented on index.bs:
```diff
@@ -349,6 +349,218 @@ One example of a feature that should be limited to secure contexts
+If the property is not inherited, then that element gets the property's initial value.
+If the property is inherited, then that element gets the parent element's computed value.
+
+The decision about whether a property should be inherited should be based largely
```
I think it should be noted that, to a first approximation, inherited or not is about whether the property affects text or boxes. The number of exceptions to that rule is fairly small, and you can go into more detail about those exceptions afterwards.

astearns replied:
If that's added I think it should go at the end of the section, as I'm not sure there's a way of stating that rule that provides as much explanation as what's currently here.

dbaron replied:
I've added a paragraph on this as the third paragraph of the section.

tabatkins commented on index.bs:
```diff
@@ -349,6 +349,218 @@ One example of a feature that should be limited to secure contexts
+Avoid values like ''yes'', ''no'', ''true'', ''false'',
+or things with more complex names that are basically equivalent to them.
+
+See [[#naming-is-hard]] for general (cross-language) advice on naming.
```
I haven't followed this link yet, but maybe note also that generic words like "mode" or "state" should be avoided in property names due to being semantically meaningless, since it's more or less true by default that a property is selecting a mode or setting a state.

dbaron replied:
Added.

tabatkins commented on index.bs:
```diff
@@ -349,6 +349,218 @@ One example of a feature that should be limited to secure contexts
+This is an abstract description of the value that is inherited when the property inherits.
+This is most important for inherited properties, but it does matter for all properties.
+
+Due to a somewhat unfortunate history,
```
I'd make this paragraph a `Note:`, since it's a diversion from the point.

tabatkins commented on index.bs:
```diff
@@ -349,6 +349,218 @@ One example of a feature that should be limited to secure contexts
+which is actually the <a>resolved value</a>,
+which can be the result of further computation starting from the computed value.
+
+The definition of a CSS property often has interactions with other CSS properties.
+Whether those interactions affect the property's computed value is a design choice,
+and it affects the behavior of inheritance.
```
```suggestion
The definition of a CSS property often has interactions with other CSS properties.
Inheritance passes down the [=computed value=], as well,
so whether or not the interaction should affect the value that descendants inherit
is an important factor in whether the interactions are handled in the [=computed value=]
or in a later value stage such as the [=used value=].
```

dbaron replied:
I rewrote this paragraph a different way from what you suggested, but incorporating a bunch of your suggestions.

tabatkins commented on index.bs:
```diff
@@ -349,6 +349,218 @@ One example of a feature that should be limited to secure contexts
+
+<a href="https://wiki.csswg.org/spec/property-dependencies">The dependencies
+that do affect computed values</a>
+are required not be circular.
```
```suggestion
are required to not be circular.
```

tabatkins commented on index.bs:
```diff
@@ -349,6 +349,218 @@ One example of a feature that should be limited to secure contexts
+so if the 'font-size' is ''20px'' then this would mean the line height is ''28px''.
+
+However, the computed value in this case is the number ''1.4'', not the length ''28px''.
+This is important because the value can inherit to elements that have a different font size:
```
```suggestion
This is important because the value can inherit to elements that have a different font size,
and preserving the "multiple of the current element's font" behavior would be broken
if it turned into a length at computed-value time:
```

tabatkins commented on index.bs:
```diff
@@ -349,6 +349,218 @@ One example of a feature that should be limited to secure contexts
+sometimes it is less complex 
+for the behavior to happen in whatever handles the computed value.
+(This complexity might be in the specification, in implementations, or both.
+Complexity in implementations might affect how fast things run for end users.)
```
Link to <https://wiki.csswg.org/spec/computed-values> for some additional details about what can and can't affect computed values?

plinss commented on index.bs:
```diff
@@ -349,6 +349,220 @@ One example of a feature that should be limited to secure contexts
+The list of values of a property should generally be chosen
+so that new values can be added.
+Avoid values like ''yes'', ''no'', ''true'', ''false'',
+or things with more complex names that are basically equivalent to them.
```
Wondering if this should be expanded to generic advice of avoiding booleans in other apis, or possibly advice as to when to use a boolean vs a string or enum.

dbaron replied:
I've moved this to #170.

atanassov commented:
It would be good to hear from @astearns as well

fantasai commented on index.bs:
```diff
@@ -349,6 +349,220 @@ One example of a feature that should be limited to secure contexts
+and the names of their values are usually adjectives (although sometimes nouns).
+
+Words in properties and values are separated by hyphens.
+Abbreviations are generally avoided.
```
We also prefer to use the base semantic word form, without additional grammatical morphemes. (This isn't a strict rule, but one that's usually followed.)

dbaron replied:
Added.

fantasai commented on index.bs:
```diff
@@ -349,6 +349,220 @@ One example of a feature that should be limited to secure contexts
+may not make sense for *applications* built using web technologies,
+it still seems like a separation worth maintaining for *documents*,
+which remain a major use of web technology.
+
```
Separation of content and presentation matters for applications also. It might be much harder to understand an app without any styling, because it requires a richer semantic vocabulary than base-level HTML does, but it does need to be understandable without CSS for non-visual UAs, and alternate styling should be possible on top of the same base markup. I don't think features designed for apps should be given a free pass on this point.

dbaron replied:
As discussed in today's teleconference, I'm going to drop this section and merge the rest.  Landing a replacement for this text is covered in #169.

astearns commented on index.bs:
```diff
@@ -349,6 +349,220 @@ One example of a feature that should be limited to secure contexts
+are related to the presentation.
+On the other hand, information about the font size makes sense to be in the style,
+since in pages that use HTML correctly
+the information in font sizes should also be obtainable from the markup
```
'the information conveyed by font sizes'

dbaron replied:
I went with "shown by".

astearns commented on index.bs:
```diff
@@ -349,6 +349,220 @@ One example of a feature that should be limited to secure contexts
+
+<h3 id="css-versus-markup">Presentation should be in CSS, content should be in HTML</h3>
+
+The separation between CSS and HTML was originally designed around the idea
```
Perhaps do 'CSS (style) and HTML (markup)' here, since you use the terms interchangeably after this? Or be consistent about using CSS and HTML instead of style and markup.

dbaron replied:
fixed

astearns commented on index.bs:
```diff
@@ -349,6 +349,220 @@ One example of a feature that should be limited to secure contexts
+Complexity in implementations might affect how fast things run for end users.)
+
+<div class="example">
+For example, the 'line-height' property accepts values that are integers,
```
s/integers/numbers/

astearns commented on index.bs:
```diff
@@ -349,6 +349,220 @@ One example of a feature that should be limited to secure contexts
+&lt;/body&gt;
+  </pre>
+
+These integer values are generally the preferred values to use for 'line-height'
```
s/integer/number/

dbaron commented:
I've addressed *most* of the feedback but not all (and marked the addressed stuff as resolved, so you can see what's left).  I also updated [the online preview](https://dbaron.github.io/design-principles/#css) and it actually worked this time because github was not broken.

dbaron commented:
As discussed in today's teleconference, I've broken out #169 and #170 for followup, and I'm going to merge this with the first subsection removed.

dbaron commented:
And definitely feel free to file followup issues if you find further issues with what has landed here.
